Enable Creation of Individual Files for Translated Rules

Mat0vu opened the pull request:
Hi everyone, 

this PR adds two new optional parameters to the sigma-cli command `sigma convert` and would close #58 
- `--output-dir` (`-od`)
- `--nesting-level` (`-nl`)

With `output_dir` it is possible to specify a directory in which all rules from the input are stored. The difference to the existing `--output` is that not all rules are written in a single file, instead within the `output_dir` there will be one file created for each translated rule.

The `nesting-level` parameter defaults to 1.  With this value, when one calls `sigma convert rules/*` with the following file hierarchy, the output_directory will contain all files on the same level. With a `nesting_level` of 2, the original structure with the parent directories `windows` and `linux` would be transferred to the output_directory.

```
rules/
├── windows/
│   └── rule_1.yml
├── linux/
│   └── rule_2.yml
```

I hope the names and descriptions of the two operators are clear. I also added two test cases. In my opinion this can be helpful for some people. Please let me know what you think and feel free to make changes.

Mat0vu committed:
Enable Creation of Individual Files for Translated Rules (#1)

* enable specifying output-dir

* add test for output creation

* add test for nesting-level

---------

Co-authored-by: Jan Aigner <jan-aigner@t-online.de>

Mat0vu committed:
improve description

thomaspatzke commented on sigma/cli/convert.py:
```diff
@@ -286,14 +332,56 @@ def convert(
+                        ctx=None, param=None, value=os.path.dirname(out_path)
+                    )
+                    with open(out_path, "w", encoding="utf-8") as f:
+                        f.write(result[index])
```
Problem: there can be more results returned by the backend than input files, e.g. if a file contains multiple rules or a rule contains multiple conditions this results in multiple queries per file.

thomaspatzke commented on sigma/cli/convert.py:
```diff
@@ -286,14 +332,56 @@ def convert(
+                    with open(out_path, "w", encoding="utf-8") as f:
+                        f.write(result[index])
+                except Exception as ex:
+                    click.echo(
```
This should raise a `click.ClickException` instead of silently dropping the error state and just output a message.

thomaspatzke commented on sigma/cli/convert.py:
```diff
@@ -21,6 +23,30 @@
 pipeline_list = list(pipeline_resolver.pipelines.keys())
 
 
+def ensure_dir_exists(ctx, param, value):
```
This seems to have problems with absolute paths including drive letters on Windows.

If I understand correctly, this replicates [`pathlib.Path.mkdir(parents=True, exist_ok=True)`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir). Preferable to using existing implementation instead of reimplementation.

thomaspatzke commented on tests/test_convert.py:
```diff
@@ -237,19 +240,93 @@ def test_convert_output_backend_option_list():
+    assert "Correlation method 'invalid' is not supported" in result.stdout
+
+
+def test_convert_write_to_output_dir():
```
Preferable to use [pyTest's `tmp_path`](https://docs.pytest.org/en/latest/how-to/tmp_path.html) as it takes care of creation and cleanup of a temporary directory structure and also has a retention of outputs for later inspection.

thomaspatzke commented:
Hi! Thanks for the pull request! I've identified an issue while code review: It's possible that this change misses to output converted queries because there can be more queries returned by the backend than input rules, see the corresponding comment for details.

Another issue I've discovered was while I tried to change to pyTest's `tmp_path` way to manage temporary outputs while testing. This has various advantages like output retention, automatic handling etc. Using `tmp_path` causes that an absolute path is provided and this caused the code to fail in creation of the output directory. It appears that `ensure_dir_exists()` has issues with drive letters in paths, which I recommend to replace anyways with `pathlib.Path.mkdir()` that basically does the same.

Mat0vu committed:
change path handling, add correlation rule test

Mat0vu commented:
Hi Thomas, 
thanks for the feedback and the tips, I´m not so familiar with pytest and didn´t know about the `tmp_path` possibility, but that´s a nice solution. I´ve changed the path handling now to use this and fixed also the `ensure_dir_exists()`. 

I can´t see any comments of you in the code review part at the moment but I also realized the issue now. I added a new test `test_convert_correlation_rule_to_output_dir` with a correlation rule and the `generate: true` attribute in the rule so that for one rule, two results are returned by the `convert` function. Currently the code would only write the first part of the result into the file. 

While for only a single translated rule this can be handled easily, I´m wondering how this is solved best in general, because if a complete directory of rules is traversed and some of them generate more than one output, how can these outputs be mapped to the corresponding paths of the input rules?

Mat0vu committed:
add uuids to correlation rule

thomaspatzke commented:
Currently I don't see a clean way to do this in a generic way for all backends that emit single queries as the link between rule and query is not maintained in the conversion process. I see two possibilities:

* If this is specific to an output format, the directory structure could be build by the backend and output as ZIP archive in a byte stream. Sigma CLI then enforces from the user that the output is written to a file.
* If this should be made generic for backends that return a list of queries then [this code in pySigma](https://github.com/SigmaHQ/pySigma/blob/30660ed4ea048aca1f5b3e5e2629991f0da78212/sigma/conversion/base.py#L165) must be changed in a way that the information about the converted rule. To my knowledge, attaching additional attributes to builtin objects as strings is not possible in Python. I think the only way to overcome this would be to add some result container object that contains this information, which would possibly break lot of stuff based on pySigma.
* Another possibility could be to track the rules in a separate array in pySigma and then allow to (optionally to maintain downward compatibility. Return this array too from `Backend.convert()`.

thomaspatzke reviewed (commented):
Implemented in #93
